hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/spark/sql/hudi/analysis/HoodieAnalysis.scala

```diff
@@ -0,0 +1,318 @@
+  * @param sparkSession
+  */
+case class HoodieAnalysis(sparkSession: SparkSession) extends Rule[LogicalPlan]
+  with SparkSqlAdapterSupport {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      // Convert to MergeIntoHoodieTableCommand
+      case m @ MergeIntoTable(target, _, _, _, _)
+        if m.resolved && isHoodieTable(target, sparkSession) =>
+          MergeIntoHoodieTableCommand(m)
+
+      // Convert to UpdateHoodieTableCommand
+      case u @ UpdateTable(table, _, _)
+        if u.resolved && isHoodieTable(table, sparkSession) =>
+          UpdateHoodieTableCommand(u)
+
+      // Convert to DeleteHoodieTableCommand
+      case d @ DeleteFromTable(table, _)
+        if d.resolved && isHoodieTable(table, sparkSession) =>
+          DeleteHoodieTableCommand(d)
+
+      // Convert to InsertIntoHoodieTableCommand
+      case l if sparkSqlAdapter.isInsertInto(l) =>
+        val (table, partition, query, overwrite, _) = sparkSqlAdapter.getInsertIntoChildren(l).get
+        table match {
+          case relation: LogicalRelation if isHoodieTable(relation, sparkSession) =>
+            new InsertIntoHoodieTableCommand(relation, query, partition, overwrite)
+          case _ =>
+            l
+        }
+      // Convert to CreateHoodieTableAsSelectCommand
+      case CreateTable(table, mode, Some(query))
+        if query.resolved && isHoodieTable(table) =>
+          CreateHoodieTableAsSelectCommand(table, mode, query)
+      case _=> plan
+    }
+  }
+}
+
+/**
+  * Rule for resolve hoodie's extended syntax or rewrite some logical plan.
+  * @param sparkSession
+  */
+case class HoodieResolveReferences(sparkSession: SparkSession) extends Rule[LogicalPlan]
+  with SparkSqlAdapterSupport {
+  private lazy val analyzer = sparkSession.sessionState.analyzer
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      // Resolve merge into
+      case MergeIntoTable(target, source, mergeCondition, matchedActions, notMatchedActions)
+        if isHoodieTable(target, sparkSession) && target.resolved && source.resolved =>
+
+        def isEmptyAssignments(assignments: Seq[Assignment]): Boolean = {
+          if (assignments.isEmpty) {
+            true
+          } else {
+            // This is a Hack for test if it is "update set *" or "insert *" for spark3.
+            // As spark3's own ResolveReference will append first five columns of the target
+            // table(which is the hoodie meta fields) to the assignments for "update set *" and
+            // "insert *", so we test if the first five assignmentFieldNames is the meta fields
+            // to judge if it is "update set *" or "insert *".
+            // We can do this because under the normal case, we should not allow to update or set
+            // the hoodie's meta field in sql statement, it is a system field, cannot set the value
+            // by user.
+            if (SPARK_VERSION.startsWith("3.")) {
```

Comment:
I see the logic in other places as well to determine spark version, would we  encapsulate a util method?

Reply:
good suggestion!